private/buf/cmd/buf/command/beta/convert/convert.go

```diff
@@ -105,50 +115,71 @@ func run(
 	if err := bufcli.ValidateErrorFormatFlag(flags.ErrorFormat, errorFormatFlagName); err != nil {
 		return err
 	}
-	messageBytes, err := io.ReadAll(container.Stdin())
-	if err != nil {
-		return err
-	}
-	if len(messageBytes) == 0 {
-		return fmt.Errorf("stdin is required as the input")
-	}
 	input, err := bufcli.GetInputValue(container, "", "")
 	if err != nil {
 		return err
 	}
-	protoSource, protoType, err := bufcli.ParseSourceAndType(ctx, input, flags.Type)
+	source, typeName, err := bufcli.ParseSourceAndType(ctx, input, flags.Type)
 	if err != nil {
 		return err
 	}
 	image, err := bufcli.NewImageForSource(
 		ctx,
 		container,
-		protoSource,
+		source,
 		flags.ErrorFormat,
-		false,
-		"",
-		nil,
-		nil,
-		false,
-		false,
+		false, // disableSymlinks
+		"",    // configOverride
+		nil,   // externalDirOrFilePaths
+		nil,   // externalExcludeDirOrFilePaths
+		false, // externalDirOrFilePathsAllowNotExist
+		false, // excludeSourceCodeInfo
 	)
 	if err != nil {
 		return err
 	}
-	message, err := bufreflect.NewMessage(ctx, image, protoType)
+	inputMessageRef, err := bufconvert.NewMessageEncodingRef(ctx, flags.Input, bufconvert.MessageEncodingBin)
+	if err != nil {
+		return fmt.Errorf("--%s: %v", outputFlagName, err)
+	}
+	message, err := bufcli.NewWireProtoEncodingReader(
+		container.Logger(),
+	).GetMessage(
+		ctx,
+		container,
+		image,
+		typeName,
+		inputMessageRef,
+	)
 	if err != nil {
 		return err
 	}
-	if err := protoencoding.NewWireUnmarshaler(nil).Unmarshal(messageBytes, message); err != nil {
+	defaultOutputEncoding, err := counterEncoding(inputMessageRef.MessageEncoding())
+	if err != nil {
 		return err
 	}
+	outputMessageRef, err := bufconvert.NewMessageEncodingRef(ctx, flags.Output, defaultOutputEncoding)
+	if err != nil {
+		return fmt.Errorf("--%s: %v", outputFlagName, err)
+	}
 	return bufcli.NewWireProtoEncodingWriter(
 		container.Logger(),
 	).PutMessage(
 		ctx,
 		container,
 		image,
 		message,
-		flags.Output,
+		outputMessageRef,
 	)
 }
+
+func counterEncoding(encoding bufconvert.MessageEncoding) (bufconvert.MessageEncoding, error) {
+	switch encoding {
+	case bufconvert.MessageEncodingBin:
+		return bufconvert.MessageEncodingJSON, nil
+	case bufconvert.MessageEncodingJSON:
+		return bufconvert.MessageEncodingBin, nil
+	default:
+		return 0, fmt.Errorf("unknown message encoding %v", encoding)
+	}
+}
```

Comment:
Let's add a comment to this. I see what you're going for with the word "counter", but it might not be immediately obvious to future readers.

Reply:
added a comment and also renamed it to `inverseEncoding` to see will it be more obvious